doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,702 @@
+// ie. what could we do or what do we want to do in the future - but are not doing now
+-->
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+In QGM, a query, as its name implies, is represented as a graph. This graph will be represented
+by a `Model` struct, that will be declared as follows:
+
+```rust
+/// A Query Graph Model instance represents a SQL query.
+struct Model {
+    /// The ID of the box representing the entry-point of the query.
+    top_box: BoxId,
+    /// All boxes in the query graph model.
+    boxes: HashMap<BoxId, Box<RefCell<QueryBox>>>,
+    /// Used for asigning unique IDs to query boxes.
+    next_box_id: Boxid,
+    /// All quantifiers in the query graph model.
+    quantifiers: HashMap<QuantifierId, Box<RefCell<Quantifier>>>,
+    /// Used for assigning unique IDs to quantifiers.
+    next_quantifier_id: QuantifierId,
+}
+```
+
+The graph has a top level box, which is the entry point of the query. In this proposal, all boxes and quantifiers
+are owned by the model and are referenced by their unique identifiers.
+
+Boxes represent high-level conceptual operators, ie. they don't correspond to any execution strategy. Each box
+has a set of input quantifiers, which describe the semantics of how the underlying boxes will be accessed.
+
+The following snippet contains the definition of the different types of operators and quantifiers. Since
+all type of boxes and quantifiers have some elements in common, both boxes and quantifiers are not directly
+represented as enum, but as struct with an enum type, containing the per-type specific members.
+
+```rust
+
+type QuantifierId = usize;
+type BoxId = usize;
+type QuantifierSet = BTreeSet<QuantifierId>;
+
+/// A semantic operator within a Query Graph.
+struct QueryBox {
+    /// uniquely identifies the box within the model
+    id: BoxId,
+    /// the type of the box
+    box_type: BoxType,
+    /// the projection of the box
+    columns: Vec<Column>,
+    /// the input quantifiers of the box
+    quantifiers: QuantifierSet,
+    /// quantifiers ranging over this box
+    ranging_quantifiers: QuantifierSet,
+    /// list of unique keys exposed by this box. Each unique key is made by
+    /// a list of column positions. Must be re-computed every time the box
+    /// is modified.
+    unique_keys: Vec<Vec<usize>>,
+    /// whether this box must enforce the uniqueness of its output, it is
+    /// guaranteed by structure of the box or it must preserve duplicated
+    /// rows from its input boxes.
+    distinct: DistinctOperation,
+}
+
+enum BoxType {
+    BaseTable(BaseTable),
+    Except,
+    Grouping(Grouping),
+    Intersect,
+    OuterJoin(OuterJoin),
+    Select(Select),
+    TableFunction(TableFunction),
+    Union,
+    Values(Values),
+}
+
+enum DistinctOperation {
```

Comment:
`DistinctOperation::Enforce` means that the operator would enforce tuple uniqueness at its output.

`SELECT DISTINCT ...` is represented as a select box with `DistinctOperation::Enforce`. If the output of a `DistinctOperation::Enforce` select is guaranteed to be distinct given the internal structure of the oeprator, a semantic rewrite make it ``DistinctOperation::Guaranteed`.

`DistinctOperation::Preserve` actually means `Preserve Duplicates`.

`DISTINCT` and `GROUP BY` have different representation in QGM. A semantic rewrite converts a `GROUP BY` box into a `DISTINCT` select box whenever possible, since select boxes are more easily merge-able.

`DISTINCT TUPLES` is a computed flag, that tells the consumer of the operator whether the output tuples are distinct or not, regardless of what was requested. This flag is used in several transforms, but at this point it is only adding confusion in the graphs.

Reply:
Exactly. I'll add that exact example in the `distinct` section.

Reply:
Added.